vcd/resource_vcd_vapp_vm.go

```diff
@@ -583,421 +591,739 @@ func vmSchemaFunc(vmType typeOfVm) map[string]*schema.Schema {
-						NetworkConnectionSection: &networkConnectionSection,
-					},
-				},
-			}
-			util.Logger.Printf("%# v", pretty.Formatter(vmParams))
-			vm, err = vdc.CreateStandaloneVMFromTemplate(&vmParams)
-			if err != nil {
-				d.SetId("")
-				return fmt.Errorf("[VM creation] error creating standalone VM %s : %s", vmName, err)
-			}
-			util.Logger.Printf("[VM create] VM after creation %# v", pretty.Formatter(vm.VM))
-			vapp, err = vm.GetParentVApp()
-			if err != nil {
-				d.SetId("")
-				return fmt.Errorf("[VM creation] error retrieving vApp from standalone VM %s : %s", vmName, err)
-			}
-			util.Logger.Printf("[VM create] vApp after creation %# v", pretty.Formatter(vapp.VApp))
-			dSet(d, "vapp_name", vapp.VApp.Name)
-		} else {
-			task, err := vapp.AddNewVMWithComputePolicy(vmName, vappTemplate, &networkConnectionSection, storageProfilePtr, sizingPolicy, acceptEulas)
-			if err != nil {
-				return fmt.Errorf("[VM creation] error adding VM: %s", err)
-			}
-			err = task.WaitTaskCompletion()
-			if err != nil {
-				return fmt.Errorf(errorCompletingTask, err)
-			}
+	// Read function is called in wrapper functions `resourceVcdVAppVmCreate` and
+	// `resourceVcdStandaloneVmCreate`
+	return nil
+}
 
-			vm, err = vapp.GetVMByName(vmName, true)
+// createVmFromTemplate is responsible for create VMs from template of two types:
+// * Standalone VMs
+// * VMs inside vApp (vApp VMs)
+//
+// Code flow has 3 layers:
+// 1. Lookup common information, required for both types of VMs (Standalone and vApp child). Things such as
+//   - Template to be used
+//   - Network adapter configuration
+//   - Storage profile configuration
+//   - VM compute policy configuration
+//
+// 2. Perform VM creation operation based on type in separate switch/case
+//   - standaloneVmType
+//   - vAppVmType
+//
+// # This part includes defining initial structures for VM and also any explicitly required operations for that type of VM
+//
+// 3. Perform additional operations which are common for both types of VMs
+//
+// Note. VM Power ON (if it wasn't disabled in HCL configuration) occurs as last step after all configuration is done.
+func createVmFromTemplate(d *schema.ResourceData, meta interface{}, vmType typeOfVm) (*govcd.VM, error) {
+	vcdClient := meta.(*VCDClient)
 
-			if err != nil {
-				d.SetId("")
-				return fmt.Errorf("[VM creation] error getting VM %s : %s", vmName, err)
-			}
-		}
+	// Step 1 - lookup common information
+	org, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return nil, fmt.Errorf(errorRetrievingOrgAndVdc, err)
+	}
 
-		var computedVmType string
-		if vapp.VApp.IsAutoNature {
-			computedVmType = string(standaloneVmType)
-		} else {
-			computedVmType = string(vappVmType)
-		}
-		// VM creation already succeeded so ID must be set
-		d.SetId(vm.VM.ID)
-		dSet(d, "vm_type", computedVmType)
+	// Lookup VM template inside vApp template - either specified by `vm_name_in_template` or the
```

Comment:
Continuing with @dataclouder suggestions 🙂 
```suggestion
	// Look up VM template inside vApp template - either specified by `vm_name_in_template` or the
```

Reply:
Fixed